File: Transport-Network/Station.java
import java.io.*;
import java.util.*;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.channels.Channel;
import java.nio.channels.DatagramChannel;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.Charset;
import java.nio.charset.CharsetEncoder;
import java.time.format.DateTimeFormatter;
import java.time.LocalDateTime;
import java.text.ParseException;
import java.text.SimpleDateFormat;

public class Station {
  private static String station_name;
  private static ArrayList<String> station_stops;
  private static Selector selector = null;
  private static HashMap<String, SocketAddress> neighbours;
  private static int udpportnumber;
  private static int tcpportnumber;
  private static DatagramChannel udpserver;

  public static void main(final String[] args) throws IOException, InterruptedException {
    station_name = args[0];
    tcpportnumber = Integer.parseInt(args[1]);
    udpportnumber = Integer.parseInt(args[2]);
    String route = "";

    neighbours = new HashMap<String, SocketAddress>();
    try {

      selector = Selector.open();
      final ServerSocketChannel tcpserver = ServerSocketChannel.open();
      tcpserver.bind(new InetSocketAddress("localhost", tcpportnumber));

      udpserver = DatagramChannel.open();
      udpserver.socket().bind(new InetSocketAddress("localhost", udpportnumber));

      tcpserver.configureBlocking(false);
      udpserver.configureBlocking(false);

      tcpserver.register(selector, SelectionKey.OP_ACCEPT);
      udpserver.register(selector, SelectionKey.OP_READ);
      Thread.sleep(1000); // sleep for 4 seconds, wait for other stations to startup to send name
      for (int i = 3; i < args.length; i++) {
        // send your stationname to each port number so they can store it
        SocketAddress stationadd = new InetSocketAddress("localhost", Integer.parseInt(args[i]));
        udp_send(udpserver, "stationname:" + station_name, stationadd);
      }

      while (true) {
        final int readychannel = selector.select();
        if (readychannel == 0)
          continue;
        final Set<SelectionKey> selectedkeys = selector.selectedKeys();
        final Iterator<SelectionKey> keyiterator = selectedkeys.iterator();
        while (keyiterator.hasNext()) {
          final SelectionKey key = keyiterator.next();
          final Channel c = key.channel();
          keyiterator.remove();
          if (!key.isValid()) {
            continue;
          }
          if (key.isReadable() && c == udpserver && key.isValid()) {

            // received message from adjacent stations
            final ByteBuffer message = ByteBuffer.allocate(1000);
            message.clear();
            SocketAddress clientAddress = udpserver.receive(message);
            message.flip();
            final int limit = message.limit();
            final byte m[] = new byte[limit];
            message.get(m, 0, limit);
            final String msg = new String(m, "US-ASCII");
            System.out.println(station_name + " received the query: " + msg);
            if (msg.contains("stationname:")) {
              // it's a station name
              String s = msg.split(":", 2)[1];
              neighbours.put(s, clientAddress);
            } else if (msg.contains("Arrival time:")) {
              // it's a finished route
              System.out.println(station_name + "Received client's route");
              route = msg;
            } else {
              // it's an unfinished route
              // check if this station is the destination
              String destination = msg.split(",")[0].split("\\s+")[2];
              if (destination.equals(station_name)) {
                String s[] = msg.split(",");
                SocketAddress firstStationadd = new InetSocketAddress(Integer.parseInt(s[0].split("\\s+")[0]));
                String arrivaltime = s[s.length - 1].split("-")[1].split("\\s+")[0];
                udp_send(udpserver, "Arrival time:" + arrivaltime + " " + msg, firstStationadd);
              } else {
                send_route(udpserver, msg);
              }
            }
          } else if (key.isAcceptable() && c == tcpserver) {
            System.out.println(station_name + "isacceptable");
            handleAccept(tcpserver, key);
          } else if (key.isReadable() && c != udpserver) {
            // receive query from tcp client
            System.out.println(station_name + "isreadable");
            handleRead(key);
          } else if (key.isWritable() && c != udpserver) {
            handleWrite(key, route);
          }
        }
      }
    } catch (final IOException e) {
      e.printStackTrace();
    }
  }

  // Finds an appropriate neighbour and sends nearest stop to that neighbour
  // Msg will be in format: "Startingport To stationB,Starting-10:00 East Station,
  // 9:15-10:30
  // tostationA, 10:40-10:50 tostationB"
  private static void send_route(final DatagramChannel mychannel, final String msg) {
    try {
      String stops[] = msg.split(",");
      // go through neighbours and look for one that hasn't been travelled to
      Iterator it = neighbours.entrySet().iterator();

      for (Map.Entry<String, SocketAddress> entry : neighbours.entrySet()) {
        String station = entry.getKey();
        SocketAddress stationaddress = entry.getValue();

        if (!travelled_before(stops, station)) {
          // current_time should be last stop's arrival time
          String current_time = stops[stops.length - 1].split("-")[1].split("\\s+")[0];
          System.out.println("Finding route for: " + station + "current time is: " + current_time);

          String earlieststop = find_earliest_stop(station, current_time);
          if (!earlieststop.equals("none")) {
            String updatedmsg = msg + "," + earlieststop;
            System.out.println(station_name + "sending the query: " + updatedmsg + "to  " + station);
            udp_send(mychannel, updatedmsg, stationaddress);
          } else {
            System.out.println("no route");
          }
        }
      }
    } catch (final IOException e) {
      e.printStackTrace();
    } catch (ParseException e) {
      e.printStackTrace();
    }
  }

  // loop through the stationlist and see if we've travelled to that station.
  private static boolean travelled_before(String[] stops, String station) {
    for (int i = 0; i < stops.length; i++) {
      if (stops[i].split("\\s+")[1].trim().equals(station)) {
        return true;
      }
    }
    return false;
  }

  // reads the timetable, and finds the earliest stop for that station with the
  // given time.
  // Return String format: stoptime-arrival arrivalstation
  private static String find_earliest_stop(String station, String time) throws ParseException {
    readfile(station_name + ".txt");
    SimpleDateFormat parser = new SimpleDateFormat("HH:mm");
    Date currenttime = parser.parse(time);
    for (int i = 0; i < station_stops.size(); i++) {
      String stopinfo[] = station_stops.get(i).split(",", 5);
      String arrivalstation = stopinfo[4].trim();
      Date stop_time = parser.parse(stopinfo[0]);
      if (arrivalstation.equals(station) && (stop_time.after(currenttime) || stop_time.equals(currenttime))) {
        // found a stop that is after the currenttime for that station
        return stopinfo[0] + "-" + stopinfo[3] + " " + arrivalstation;
      }
    }
    return "none";
  }

  // Sends a string to a udp server after encoding it
  private static void udp_send(final DatagramChannel mychannel, String message, SocketAddress address)
      throws CharacterCodingException {
    try {
      CharsetEncoder encoder = Charset.forName("US-ASCII").newEncoder();
      ByteBuffer udpmessage = ByteBuffer.allocate(1000);
      udpmessage.clear();
      udpmessage = encoder.encode(CharBuffer.wrap(message));
      mychannel.send(udpmessage, address);
    } catch (IOException e) {
      System.out.println("IO exception");
      e.printStackTrace();
    }
  }

  private static void readfile(final String filename) {
    try {
      station_stops = new ArrayList<String>();
      // first time reading timetable or updating timetable
      File timetable_file = new File(filename);
      final BufferedReader br = new BufferedReader(new FileReader(timetable_file));
      String st;
      br.readLine(); // ignore first line
      while ((st = br.readLine()) != null) {
        station_stops.add(st); // put the station stops in an array
      }
    } catch (final FileNotFoundException e) {
      System.out.println("File not found error");
      e.printStackTrace();
    } catch (final IOException e) {
      System.out.println("File not found error");
      e.printStackTrace();
    }
  }

  private static String get_current_time() {
    DateTimeFormatter dtf = DateTimeFormatter.ofPattern("HH:mm");
    LocalDateTime now = LocalDateTime.now();
    System.out.println(dtf.format(now));
    return dtf.format(now);
  }

  private static void handleAccept(final ServerSocketChannel mySocket, final SelectionKey key) throws IOException {
    // Accept the connection and set non-blocking mode
    final SocketChannel client = mySocket.accept();
    if (client != null) {
      client.configureBlocking(false);
      System.out.println("Connection Accepted...");
      // Register that client is reading this channel
      client.register(selector, SelectionKey.OP_READ);
      final String response = "HTTP/1.1 200 OK\nContent-Type: text/html\nConnection: Closed\n\n<html><body><h1>Finding your requested route..</h1></body></html>";
      CharsetEncoder encoder = Charset.forName("US-ASCII").newEncoder();
      final ByteBuffer responsebuf = encoder.encode(CharBuffer.wrap(response));
      client.write(responsebuf);
    }
  }

  private static void handleRead(final SelectionKey key) throws IOException {
    System.out.println("Reading...");
    // create a ServerSocketChannel to read the request
    final SocketChannel client = (SocketChannel) key.channel();
    // Create buffer to read data
    final ByteBuffer buffer = ByteBuffer.allocate(400);
    client.read(buffer);
    // Parse data from buffer to String
    final String data = new String(buffer.array()).trim();
    if (data.length() > 0 && data.contains("GET /?to=")) {
      System.out.println("Received message: " + data);
      final String s[] = data.split("=", 2);
      String destination = s[1].substring(0, s[1].indexOf(" "));
      String msg = udpportnumber + " To " + destination + "," + "Starting-" + get_current_time() + " " + station_name;
      send_route(udpserver, msg);
      client.register(selector, SelectionKey.OP_WRITE);
    } else {
      client.close();
      System.out.println("connection closed");
    }
  }

  private static void handleWrite(final SelectionKey key, String route) throws CharacterCodingException {
    if (route != "") {
      final SocketChannel client = (SocketChannel) key.channel();
      CharsetEncoder encoder = Charset.forName("US-ASCII").newEncoder();
      final ByteBuffer responsebuf = encoder.encode(CharBuffer.wrap(route));
      try {
        client.write(responsebuf);
        client.close();
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
  }
}

File: Transport-Network/Station.py
import select
import socket
import sys
import Queue
import SocketServer
from datetime import datetime
import time

station_name = sys.argv[1]
tcpportnumber = int(sys.argv[2])
udpportnumber = int(sys.argv[3])
adjacent_stations = sys.argv[4:]
inputs = []
outputs = []
neighbours = {} #dictionary Stationname: portnumber
completedroute = ""
routes_sent = 0
noroutes_received = 0

print(station_name + "setting up tcp server")
tcpserver = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
tcpserver.setblocking(0)
tcpserver.bind(("localhost", tcpportnumber))
tcpserver.listen(1)
inputs.append(tcpserver)

print("starting udp server")
udpserver = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
udpserver.setblocking(0)
udpserver.bind(("localhost", udpportnumber))
inputs.append(udpserver)

time.sleep(1) # wait for other stations to start before sending name
#send your station name to adjacent ports
for station in adjacent_stations:
  myname = "stationname:" + station_name
  udpserver.sendto(myname, ("localhost", int(station)))

def tcphandle_read(tcpclient):
  data = tcpclient.recv(1024)
  print ('received "%s" from %s' % (data, tcpclient.getpeername()))
  if "GET /?to=" in data:
    s = data.split("=", 2)
    destination = s[1][0:s[1].index(" ")]
    print("destination is :"+destination)
    msg = str(udpportnumber) + " To " + destination + "," + "Starting-" + str(get_current_time()) + " "+ station_name;
    print(msg)
    send_route(udpserver, msg)
    outputs.append(tcpclient)
  else:
    tcpclient.close()

def udphandle_read(udpserver):
  global completedroute
  global routes_sent
  global noroutes_received
  data,address = udpserver.recvfrom(1024)
  if "stationname:" in data:
    stationname = data.split(":")[1]
    neighbours[stationname] = address
  elif "Arrival time:" in data:
    print("Received client's route" + data)
    completedroute = data
  elif "No route:" in data:
    #data format - No route: 4004 To station a, starting-10:00 station b, 10:00-10:30 station c
    noroutes_received = noroutes_received + 1
    if noroutes_received == routes_sent:
      #every route this station sent out has returned with no route
      #check if this station is the first station
      print(data)
      firststation_add = int(data.split(",")[0].split()[2])
      if udpportnumber == firststation_add:
        print("It's the same")
        completedroute = "No route found"
        routes_sent = 0
        noroutes_received = 0
      else:
        send_noroute(udpserver, data)
  else:
    msg = data.split(",")
    #unfinished route, check if this station is the destination
    destination = msg[0].split()[2]
    if destination == station_name:
      firststation_add = int(msg[0].split()[0])
      arrival_time = msg[len(msg)-1].split("-")[1].split()[0]
      udpserver.sendto("Arrival time:" + arrival_time +" "+ data, ("localhost", firststation_add))
    else:
      send_route(udpserver, data)

def handle_write():
  if completedroute != "":
    print("Sending route to client")
    s.send("HTTP/1.1 200 OK\nContent-Type: text/html\nConnection: Closed\n\n<html><body><h1>Finding your requested route..</h1></body></html>");
    s.send(completedroute)
    outputs.remove(s)
    s.close
  else :
    pass

def send_route(socket, msg):
  found_route = False
  global routes_sent
  global noroutes_received
  stops = msg.split(",")[1:]
  for station in neighbours:
    if not has_travelled(station, stops):
      current_time = stops[len(stops)-1].split("-")[1].split()[0]
      earlieststop = find_earliest_stop(station, current_time)
      print(earlieststop)
      if not earlieststop == "none":
        updatedmsg = msg + "," + earlieststop
        socket.sendto(updatedmsg, (neighbours[station]))
        routes_sent = routes_sent + 1
        found_route = True
        print("Sending route" + updatedmsg + "to " + str(neighbours[station]))
  if found_route == False:
    send_noroute(socket, msg)
    routes_sent = routes_sent + 1
  
def find_earliest_stop(adjstation, time):
  print("finding earliest stop for:" + adjstation)
  with open(station_name + ".txt") as timetable:
    stops = timetable.readlines()
  current_time = datetime.strptime(time, "%H:%M")
  for stop in stops[1:]:
    stopinfo = stop.split(",")
    arrivalstation = stopinfo[4].strip()
    stop_time = datetime.strptime(stopinfo[0],"%H:%M")
    arrival_time = stopinfo[3]
    if arrivalstation == adjstation and (stop_time > current_time or stop_time == current_time):
      return stopinfo[0] + "-" + arrival_time + " " + arrivalstation
  return "none"
  
#given A string with the format- No route:4004 To station a, starting-10:00 station b, 10:00-10:30 station c
#find the station before the current one and pass the route to that station
def send_noroute(socket, route):
  routes = route.split(",")[1:]
  if len(routes) == 1:
    #starting station didn't find a route
    global completedroute
    completedroute = "No route found"
    return
  if not "No route: " in route:
    route = "No route: " + route
  for stop in routes:
    currentstation = stop.split()[1]
    if currentstation == station_name:
      i = routes.index(stop)
      previous_station = routes[i-1].split()[1]
      print(previous_station)
      socket.sendto(route, (neighbours[previous_station]))

def has_travelled(station, stops):
  for stop in stops:
    currentstation = stop.split()[1]
    if station == currentstation:
      return True
  return False

def get_current_time():
  now = datetime.now()
  current_time = now.strftime("%H:%M")
  return current_time

#selector loop
while True: 
  readable, writable, exceptional = select.select(inputs, outputs, inputs)
  for s in readable:
    if s is tcpserver:
      print("tcp handling accept")
      (connection, client_address) = tcpserver.accept()
      connection.setblocking(0)
      inputs.append(connection) #add it to be able to read
    elif s is udpserver:
      udphandle_read(udpserver)
    else:
      tcphandle_read(s)
      
  for s in writable:
    handle_write()
    

